packages/node_modules/@webex/plugin-meetings/src/roap/turnDiscovery.ts

```diff
@@ -0,0 +1,238 @@
+import {Defer} from '@webex/common';
```

Comment:
This just looks like a promise , Do you like the name Defer or we need to change it to WaitForPromise or something 

This code is good but is there a general pattern which we can use in the future !?

Reply:
Yeah I personally prefer a name like "ControlledPromise", but "defer" seems to be a more commonly known name for this so maybe we should keep it. Defer is also sometimes referred to as an anti-pattern, mainly because of the issue if there is any error happening that prevents the `defer.resolve()`, then you might be stuck forever waiting for it. In this case we have a timeout in `waitForTurnDiscoveryResponse()` so it's not a problem, but it's something to be aware of when using Defer in general.
I think overall the more general pattern to be used in scenarios like this one is a state machine, but because here it's very simple - we only wait for one message, then I thought that just using a single defer here will make the code easier to read than having to define a state machine. It allows me to just write 
```
this.sendRoapTurnDiscoveryRequest(meeting, isReconnecting)
      .then(() => this.waitForTurnDiscoveryResponse())
      .then(() => this.sendRoapOK(meeting))
```
in `doTurnDiscovery()` and it's immediately clear how this whole turn discovery sequence works.